crates/noirc_evaluator/src/ssa/optim.rs

```diff
@@ -229,75 +155,82 @@ pub fn find_similar_instruction_with_multiple_arguments(
 
 pub fn find_similar_cast(
     igen: &SsaContext,
-    lhs: NodeId,
+    operator: &Operation,
     res_type: node::ObjectType,
     prev_ins: &VecDeque<NodeId>,
 ) -> Option<NodeId> {
     for iter in prev_ins {
         if let Some(ins) = igen.try_get_instruction(*iter) {
-            if ins.lhs == lhs && ins.res_type == res_type {
+            if &ins.operator == operator && ins.res_type == res_type {
```

Comment:
Same remark, let's check on the argument and type as before so we are not affected by whatever is added to the operation.

Reply:
similarly, this can be refactored in another PR.